src/PluginCreator.ts

```diff
@@ -0,0 +1,208 @@
+import * as ts from 'typescript';
+import * as resolve from 'resolve';
+import * as path from 'path';
+import compareVersions from 'compare-versions';
+import { parseConfigFileTextToJson } from 'typescript/lib/tsserverlibrary';
+
+export type FactoryType = 'ls' | 'program' | 'opts' | 'checker';
+
+export type PluginConfig = {
+    name?: string;
+    transform?: string;
+    type: FactoryType;
+    after?: boolean;
+    before?: boolean;
+    afterDeclaration?: boolean;
+};
+
+export type TsTransformerFactory = ts.TransformerFactory<ts.SourceFile>;
+
+export type TransformerPlugin = {
+    before?: TsTransformerFactory;
+    after?: TsTransformerFactory;
+    afterDeclaration?: TsTransformerFactory;
+};
+
+export type FactoryRet = TransformerPlugin | TsTransformerFactory;
+export type LSPattern = (ls: ts.LanguageService, config?: PluginConfig) => FactoryRet;
+export type ProgramPattern = (program: ts.Program, config?: PluginConfig) => FactoryRet;
+export type CompilerOptionsPattern = (opts: ts.CompilerOptions, config?: PluginConfig) => FactoryRet;
+export type TypeCheckerPattern = (opts: ts.TypeChecker, config?: PluginConfig) => FactoryRet;
+export type DefaultPattern = (context: ts.TransformationContext, program?: ts.Program) => ts.Transformer<ts.SourceFile>;
+export type PluginFactory = LSPattern | ProgramPattern | CompilerOptionsPattern | TypeCheckerPattern | DefaultPattern;
+
+function patchEmitFiles(host: any): ts.TransformerFactory<ts.SourceFile>[] {
+    const oldEmitFiles = host.emitFiles;
+    if (oldEmitFiles.__patched instanceof Array) {
+        oldEmitFiles.__patched.length = 0;
+        return oldEmitFiles.__patched;
+    }
+
```

Comment:
Can you refactor this code to be more clear, I really don't understand what happens here
Why there is a `oldEmitFiles.__patched`?

Reply:
createTransformers can be called many times with same ts instance.
Without __patched, on second createTransformers call -  new transformers will be ignored and used from oldCreateTransformers.